cmd/credentialUtil.go

```diff
@@ -135,17 +135,62 @@ func GetOAuthTokenManagerInstance() (*common.UserOAuthTokenManager, error) {
 // 4. If there is OAuth token info passed from env var, indicating using token credential. (Note: this is only for testing)
 // 5. Otherwise use anonymous credential.
 // The implementation logic follows above rule, and adjusts sequence to save web request(for verifying public resource).
-func getBlobCredentialType(ctx context.Context, blobResourceURL string, canBePublic bool, standaloneSAS bool, cpkOptions common.CpkOptions) (common.CredentialType, bool, error) {
+func getBlobCredentialType(ctx context.Context, blobResourceURL string, canBePublic bool, standaloneSAS string, cpkOptions common.CpkOptions) (common.CredentialType, bool, error) {
 	resourceURL, err := url.Parse(blobResourceURL)
 
 	if err != nil {
 		return common.ECredentialType.Unknown(), false, errors.New("provided blob resource string is not in URL format")
 	}
 
+	if standaloneSAS != "" {
+		resourceURL.RawQuery = standaloneSAS
+	}
+
 	sas := azblob.NewBlobURLParts(*resourceURL).SAS
+	isMDAccount := strings.HasPrefix(resourceURL.Host, "md-")
+	canBePublic = canBePublic && !isMDAccount // MD accounts cannot be public.
 
 	// If SAS existed, return anonymous credential type.
-	if isSASExisted := sas.Signature() != ""; isSASExisted || standaloneSAS {
+	if isSASExisted := sas.Signature() != ""; isSASExisted {
+		if isMDAccount {
+			// Ping the account anyway, and discern if we need OAuth.
+			p := azblob.NewPipeline(
+				azblob.NewAnonymousCredential(),
+				azblob.PipelineOptions{
+					Retry: azblob.RetryOptions{
+						Policy:        azblob.RetryPolicyExponential,
+						MaxTries:      ste.UploadMaxTries,
+						TryTimeout:    ste.UploadTryTimeout,
+						RetryDelay:    ste.UploadRetryDelay,
+						MaxRetryDelay: ste.UploadMaxRetryDelay,
+					},
+					RequestLog: azblob.RequestLogOptions{
+						SyslogDisabled: common.IsForceLoggingDisabled(),
+					},
+				})
+
+			clientProvidedKey := azblob.ClientProvidedKeyOptions{}
+			if cpkOptions.IsSourceEncrypted {
+				clientProvidedKey = common.GetClientProvidedKey(cpkOptions)
+			}
+
+			bURL := azblob.NewBlobURL(*resourceURL, p)
+			_, err := bURL.GetProperties(ctx, azblob.BlobAccessConditions{}, clientProvidedKey)
+
+			if err != nil {
+				if stgErr, ok := err.(azblob.StorageError); ok {
+					if httpResp := stgErr.Response(); httpResp.StatusCode == 401 {
```

Comment:
Nit: Please add comments in here to explain what is happening in this segment of code.